### DISCUSSION TITLE
Shark model fitting with dives and location quality

### DISCUSSION THREAD

**Discussion by meg-abytes** [bmcclintock/momentuHMM#122] [Q&A]
I am currently working through the package vignette and have looked at several examples- I am **very** new to state space modelling and am trying to wrap my head around the best way to proceed with the following model for shark data. 

What I have:
Irregular time spaced Argos location data with Quality fixes (3,2,1,0,A,B) 
High resolution dive data (every 10 seconds depth and temperature)

What I have done so far:
used `prepData`, `crawlwrap `and `fitHMM ` to fit a model for the argos location data (not accounting for location quality)- see code below:

```
rawdata %>% distinct(Latitude,Longitude, .keep_all= TRUE) -> ws_data
ws_data <- ws_data[,c("Ptt","date.time","Longitude","Latitude","Quality")]
colnames(ws_data) <- c("ID","time", "x", "y","lc") #name x and y for prepdata
ws_data$lc <- as.factor(ws_data$lc)
ws_data$time <- format(round(ws_data$time, units = 'hour'), format = "%Y-%m-%d %H:%M:%S")
ws_data$time <- as.POSIXct(ws_data$time, format = c("%Y-%m-%d %H:%M:%S"), tz = "UTC")


##prep data
hmmdata <- prepData(ws_data)
plot(hmmdata,ask=FALSE)
#crawl wrap
crwOut <-crawlWrap(obsData = ws_data, retryFits=20, timeStep = "hour") #fits CTCRW model and predict hourly locations
plot(crwOut,ask=FALSE)
crwOut$crwFits
#get acf
acf(hmmdata$step[!is.na(hmmdata$step)],lag.max=300)

## label states
stateNames <- c("forage","travel")
# distributions for observation processes
dist = list(step = "gamma", angle = "wrpcauchy")

#choosing parameters#
stateswitch <- 0.4 # "stateswitch" value based on histogram
#filter for steps less than state switch value
hmmdata %>%
  filter(step <= stateswitch) -> hmmdataa 
hmmdata %>%
  filter(step > stateswitch) -> hmmdatab #filter for steps greater than state switch value
#calculate mean for steps of <or> each state switch value
st1.mean<-mean(hmmdataa$step, na.rm= TRUE) 
st1.sd <- sd(hmmdataa$step) 
st2.mean<-mean(hmmdatab$step, na.rm= TRUE)
st2.sd <- sd(hmmdatab$step)
# Display first rows of data set
head(hmmdata)

h <- hist(hmmdata$step, breaks = 50) #display histograms of steps to choose state 1, 2 (foraging, traveling)
h$breaks
summary(hmmdata$step)
## Starting values for the step length parameters
stepMean0 <- c(st1.mean, st2.mean) # initial means (one for each state)
stepSD0 <- c(st1.sd, st2.sd) # initial standard deviations (one for each state)
stepPar0 <- c(stepMean0, stepSD0)
## Indices of steps of length zero
whichzero <- which(hmmdata$step == 0)
## Proportion of steps of length zero in the data set
length(whichzero)/nrow(hmmdata)
## Plot histogram of turning angles
hist(hmmdata$angle, breaks = seq(-pi, pi, length = 15))

##Starting values for the turning angle parameters #
st1.amean<- mean(hmmdataa$angle, na.rm = TRUE)
st2.amean<-mean(hmmdatab$angle, na.rm= TRUE)

##initial means (one for each state) 
angleMean0 <- c(st1.amean, st2.amean) 
angleCon0 <- c(0.5,3) # initial concentrations (one for each state) 
anglePar0 <- c(angleMean0, angleCon0)

## initial parameters
Par0_m1 <- list(step=stepPar0, angle=c(0.3,0.7)) #chose initial parameters from elephant example- don't know how to get them?
#get model
m1 <- fitHMM(data = hmmdata, nbStates = 2, dist = dist, Par0 = Par0_m1,
             estAngleMean = list(angle=FALSE), stateNames = stateNames)
Par = getPar(m1) #see initial parameters
```

What I am trying to do:
1) fit a model that includes location error data (similar to the turtle example vignette)
2) fit another model based on diving behaviour to indicate foraging (similar to nfs example vignette)

I have so much information from the tags that could likely be applied to diving related foraging, however, since I am so new to this I have kept it simple by finding the number of U-shaped dives per hour (and created a dataframe similar to the nfs example (with ID, time, dive) showing how many dives are occurring per hour of the tag duration.

following the nfs example I have merged them using crawlmerge:

```
predTimes <- seq(max(foragedives$time[1],rawData$time[1]),min(foragedives$time[nrow(foragedives)],rawData$time[nrow(rawData)])+60*60,"hour")

# merge crwData object with forage dive data
crwOutmerge <- crawlMerge(crwOut,foragedives,"time")
```

I then followed the rest of the example as is (lines 27-63) which may not be correct for what I am trying to do?

I then get stuck on the Par = Par:

```
Par0 <- getParDM(nbStates = nbStates, dist = dist,
                 Par = Par, DM = DM, workBounds = workBounds,
                 estAngleMean = list(angle = FALSE))
```

with the following error message: _`Error in getParDM.default(nbStates = nbStates, dist = dist, Par = Par,  : 
  stepangledive is missing in 'Par`_

I know that I need to include the dives in the model somehow and choose the initial parameters for them, but I am unsure how to do this with my data. I have read through the vignette for moveHMM for choosing initial parameters, but am unsure how to implement that to my diving data. I have also read through the examples provided for the momentuHMM vignette and see that there are some examples that use dive data with no location data (in the harbour porpoise example). However, because sharks do not need to surface, I am not sure how to implement this in a model in a similar way. (I have similar dive summary information from the harbor porpoise example that I could also use instead of divesperhour idea of the nfs example if I knew how)

My specific questions are:
1) How do I incorporate location quality data in my model?
2) How do I also incorporate dive data in my model? (does it need to be done separately or with the first model fit?)
3) I am still unsure about choosing initial parameters- how do I choose them based on my graphs produced? The histograms I get are very different than the examples provided.
4) How do I choose the starting parameters for dives?
4) In the elephant example, where do the angle values come from? (0.3 and 0.7) and how do I calculate them for my data. I played around with values and found that I can only choose numbers between 0-1 which makes sense based on what I have read, but I am not sure how to calculate them specifically for my data.

I have a very large, complete dataset that I would love to incorporate a model to determine foraging efforts, but, am not sure how to proceed. Any help from anyone who has done something similar would be very much appreciated!!!! 

Here are some of my model outputs based only on my location data with no quality of fixes included:
![image](https://user-images.githubusercontent.com/82802382/153335977-6bf7150b-3257-4160-b820-c25872030f7c.png)

![image](https://user-images.githubusercontent.com/82802382/153336024-3df6ab22-a67e-4830-9150-54ed932116f9.png)

![image](https://user-images.githubusercontent.com/82802382/153336039-1febd5e4-eb55-4e60-a420-d05518689e6e.png)

![image](https://user-images.githubusercontent.com/82802382/153336109-9438130e-9c2c-412f-acd5-d346fab90a9f.png)

![image](https://user-images.githubusercontent.com/82802382/153336124-1ba74869-ed1d-487f-8e12-a3bd0768fef3.png)

![image](https://user-images.githubusercontent.com/82802382/153336146-1ef57e8a-9ce4-4b94-8344-b9665eb6f605.png)

```
Values calculated from the moveHMM starting values vignette:

Value of the maximum log-likelihood: -1774.898 

Step length parameters:
----------------------
      state 1  state 2
mean 2.377139 33.50524
sd   2.552155 53.49973

Turning angle parameters:
------------------------
                state 1    state 2
mean          2.4298463 -0.2189463
concentration 0.0322818  0.3962030

Regression coeffs for the transition probabilities:
--------------------------------------------------
           1 -> 2    2 -> 1
intercept -3.4394 -1.614943

Transition probability matrix:
-----------------------------
          [,1]       [,2]
[1,] 0.9689134 0.03108657
[2,] 0.1659034 0.83409659

Initial distribution:
--------------------
[1] 9.999995e-01 4.639994e-07
```

**Comment by bmcclintock** [comment]
This is a lot to ask and guidance for all of your questions can be found in the `momentuHMM` and `moveHMM` vignettes, references therein, and in GitHub [Discussions](https://github.com/bmcclintock/momentuHMM/discussions). Before delving further into an analysis with your own data, I'd suggest taking a really deep dive into these resources until you have a firm understanding of these methods, including working through all of the [examples](https://github.com/bmcclintock/momentuHMM/tree/master/vignettes/examples) in the vignette. Specifically:

> 1. How do I incorporate location quality data in my model?

`crawlWrap` is a wrapper for the `crawl` functions `crwMLE`, `crwSimulator`, and `crwPredict`. Consult the help files for these functions as well as the many examples in the `momentuHMM` [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf) (and references therein). Search for "crawl" in the GitHub [Discussions](https://github.com/bmcclintock/momentuHMM/discussions). There is also a lot of great `crawl` workshop material [here](https://jmlondon.github.io/crawl-workshop/).

> 2. How do I also incorporate dive data in my model? (does it need to be done separately or with the first model fit?)

See the northern fur seal and harbor seal examples in the momentuHMM [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf). See also [McClintock 2017](https://doi.org/10.1007/s13253-017-0285-6).

> 3. I am still unsure about choosing initial parameters- how do I choose them based on my graphs produced? The histograms I get are very different than the examples provided.

See this `moveHMM` [vignette](https://cran.r-project.org/web/packages/moveHMM/vignettes/moveHMM-starting-values.pdf). See #55, #84, and search for "initial values" in the GitHub [Discussions](https://github.com/bmcclintock/momentuHMM/discussions). 

> 4. How do I choose the starting parameters for dives?

See above.

> 5. In the elephant example, where do the angle values come from? (0.3 and 0.7) and how do I calculate them for my data. I played around with values and found that I can only choose numbers between 0-1 which makes sense based on what I have read, but I am not sure how to calculate them specifically for my data.

See above.
